src/com/xilinx/rapidwright/router/VersalClockRouting.java

```diff
@@ -0,0 +1,782 @@
+        int watchDog = 10000000;
+        RouteNode centroidHRouteNode = null;
+        Set<Node> visited = new HashSet<>();
+
+        // In Vivado solutions, we can always find the pattern:
+        // ... -> NODE_GLOBAL_GCLK -> NODE_GLOBAL_VROUTE -> NODE_GLOBAL_VDISTR_LVL2 -> ...
+        // and this is how we locate the VROUTE node
+
+        while (!q.isEmpty()) {
+            RouteNode curr = q.poll();
+            Node currNode = Node.getNode(curr);
+            RouteNode parent = curr.getParent();
+            for (Node downhill : currNode.getAllDownhillNodes()) {
+                if (parent != null) {
+                    Node parentNode = Node.getNode(parent);
+                    if (parentNode.getIntentCode() == IntentCode.NODE_GLOBAL_VROUTE &&
+                        currNode.getIntentCode() == IntentCode.NODE_GLOBAL_HROUTE_HSR) {
+                        // Disallow ability to go from VROUTE back to HROUTE
+                        continue;
+                    }
+                    if (downhill.getIntentCode()     == IntentCode.NODE_GLOBAL_VDISTR_LVL2 &&
+                       currNode.getIntentCode()   == IntentCode.NODE_GLOBAL_GCLK &&
+                       parentNode.getIntentCode() == IntentCode.NODE_GLOBAL_VROUTE &&
+                       clockRegion.equals(currNode.getTile().getClockRegion()) &&
+                       clockRegion.equals(parentNode.getTile().getClockRegion()) &&
+                       parentNode.getWireName().contains("BOT")) {
+                        if (findCentroidHroute) {
+                            centroidHRouteNode = curr.getParent();
+                            while (centroidHRouteNode.getIntentCode() != IntentCode.NODE_GLOBAL_HROUTE_HSR) {
+                                centroidHRouteNode = centroidHRouteNode.getParent();
+                            }
+                            clk.getPIPs().addAll(centroidHRouteNode.getPIPsBackToSourceByNodes());
+                            return centroidHRouteNode;
+                        }
+                        // assign PIPs based on which RouteNode returned, instead of curr
+                        clk.getPIPs().addAll(parent.getPIPsBackToSourceByNodes());
+                        return parent;
+                    }
+                }
+
+                // Only using routing lines to get to centroid
+                if (!allowedIntentCodes.contains(downhill.getIntentCode())) {
+                    continue;
+                }
+                if (!findCentroidHroute && downhill.getIntentCode() == IntentCode.NODE_GLOBAL_HROUTE_HSR) {
+                    continue;
+                }
+                if (visited.contains(downhill)) continue;
+                RouteNode rn = new RouteNode(downhill.getTile(), downhill.getWireIndex(), curr, curr.getLevel()+1);
+                
+                // The clockRegion.getApproximateCenter() may return an INVALID_* tile with huge coordinates.
```

Comment:
Can you provide some more information (better yet, create a new PR with a test)

Reply:
Sorry that I wrongly compared their coordinates with the number of rows and columns of xcvu3p. I made a test for  clockRegion.getApproximateCenter() and the coordinates of returned tiles works correctly.

Reply:
Then is this comment outdated and can be removed?

Reply:
> Then is this comment outdated and can be removed?

Yes

Reply:
Should we go back to using `getApproximateCenter()` then?

Reply:
(I went ahead and changed it, and it seems to continue to work)